ratik/src/main/java/org/opencadc/ratik/InventoryValidator.java

```diff
@@ -432,22 +435,38 @@ String buildRemoteQuery(final String bucket)
         return query.toString();
     }
 
+    /**
+     * Method that can be overridden in tests to insert Artifact's after the iterator queries.
+     */
+    void testAction() { }
```

Comment:
Is this `testAction()` here so you can simulate the "artifact changed while iterating" that gets detected after acquiring a lock? It is admittedly hard to test that code handles race conditions (two different threads/processes are changing state and the actual sequence of events is a "race"). I haven't tried to do it before because there are so many unknowns and I'd normally rely on careful code review/analysis.... 

Methods like this (and in general odd/unnatural code to enable testing) make me nervous. Plus you can't really know if any action you take will or will not be visible in the iterator (eg you add a row to the db and it could show up in the iterator or not depending on all kinds of subtle things like row fetch size, query plan, db locking, etc).

I haven't looked at the test changes to make use of this, but I would expect this to be fragile at best, while the "lock && get" pattern in code is easy to look for. 

So, do you think this is really worth it? Are the tests going to hold up? Are the going to work and be tests easy to grok in six months?

Reply:
Yes, the `testAction()` method is solely for a single test case where after the local and remote iterators have been created an Artifact is put into the local site.
I agree it's a hack for a single test case, but it was the only way I could come up with to test that discrepancy.
Is it worth it? For a single test case probably not. Maybe if it was useful for multiple cases we could pretend not to see it.
Will they work in 6 months? Depends on if they are maintained or not. Easy to understand comes from documentation and comments in the tests. I think the docs explains the each test case well enough. More comments in the tests about what they are doing and why could be helpful down the road. I'll go through and add comments to the tests to explain them better.